File: Task 2/Java/app/src/main/java/dev/ogabek/java/network/VolleyHttp.java
package dev.ogabek.java.network;

import androidx.annotation.Nullable;

import com.android.volley.AuthFailureError;
import com.android.volley.Request;
import com.android.volley.toolbox.StringRequest;

import org.json.JSONObject;

import java.util.HashMap;
import java.util.Map;

import dev.ogabek.java.MyApplication;
import dev.ogabek.java.helper.Logger;
import dev.ogabek.java.model.Employee;

public class VolleyHttp {

    private static final String TAG = VolleyHttp.class.toString();
    public static final boolean IS_TESTER = true;
    private static final String SERVER_DEVELOPMENT = "http://dummy.restapiexample.com/api/v1/";
    private static final String SERVER_PRODUCTION = "http://dummy.restapiexample.com/api/v1/";

    static String server(String url) {
        if (IS_TESTER) {
            return SERVER_DEVELOPMENT + url;
        } else {
            return SERVER_PRODUCTION + url;
        }
    }

    static Map<String, String> header() {
        Map<String, String> header = new HashMap<>();
        header.put("Content-type", "application/json; charset=UTF-8");
        return header;
    }

    // Request Methods

    public static void get(String api, HashMap<String, String> params, VolleyHandler volleyHandler) {
        StringRequest request = new StringRequest(
                Request.Method.GET, server(api),
                response -> {
                    Logger.d(TAG, response);
                    volleyHandler.onSuccess(response);
                },
                error -> {
                    Logger.d(TAG, error.toString());
                    volleyHandler.onError(error.getMessage());
                }) {
            @Nullable
            @Override
            protected Map<String, String> getParams() {
                return params;
            }
        };
        MyApplication.instance.addToRequestQueue(request);
    }

    public static void post(String api, HashMap<String, String> body, VolleyHandler volleyHandler) {
        StringRequest request = new StringRequest(
                Request.Method.POST, server(api),
                response -> {
                    Logger.d(TAG, response);
                    volleyHandler.onSuccess(response);
                },
                error -> {
                    Logger.d(TAG, error.toString());
                    volleyHandler.onError(error.toString());
                }) {
            @Override
            public byte[] getBody() {
                return new JSONObject((Map<String, String>) body).toString().getBytes();
            }

            @Override
            public Map<String, String> getHeaders() {
                return header();
            }
        };
        MyApplication.instance.addToRequestQueue(request);
    }

    public static void  put(String api, HashMap<String, String> body, VolleyHandler volleyHandler) {
        StringRequest request = new StringRequest(
                Request.Method.PUT, server(api),
                response -> {
                    Logger.d(TAG, response);
                    volleyHandler.onSuccess(response);
                },
                error -> {
                    Logger.d(TAG, error.toString());
                    volleyHandler.onError(error.toString());
                }) {
            @Override
            public byte[] getBody() {
                return new JSONObject((Map<String, String>) body).toString().getBytes();
            }

            @Override
            public Map<String, String> getHeaders() {
                return header();
            }
        };
        MyApplication.instance.addToRequestQueue(request);
    }

    public static void delete(String api, VolleyHandler volleyHandler) {
        StringRequest request = new StringRequest(
                Request.Method.DELETE, server(api),
                response -> {
                    Logger.d(TAG, response);
                    volleyHandler.onSuccess(response);
                },
                error -> {
                    Logger.d(TAG, error.toString());
                    volleyHandler.onError(error.toString());
                }) {
        };
        MyApplication.instance.addToRequestQueue(request);
    }

    // Request API's

    public static final String API_LIST_EMPLOYEES = "employees";
    public static final String API_SINGLE_EMPLOYEE = "employee/"; //id
    public static final String API_CREATE_EMPLOYEE = "create";
    public static final String API_UPDATE_EMPLOYEE = "update/"; //id
    public static final String API_DELETE_EMPLOYEE = "delete/"; //id

    // Request Params

    public static HashMap<String, String> paramsEmpty() {
        return new HashMap<>();
    }

    public static HashMap<String, String> paramsCreate(Employee employee) {
        HashMap<String, String> params = new HashMap<>();
        params.put("employee_name", employee.getEmployee_name());
        params.put("employee_salary", employee.getEmployee_salary());
        params.put("employee_age", String.valueOf(employee.getEmployee_age()));
        params.put("profile_image", employee.getProfile_image());
        return params;
    }

    public static HashMap<String, String> paramsUpdate(Employee employee) {
        HashMap<String, String> params = new HashMap<>();
        params.put("id", String.valueOf(employee.getId()));
        params.put("employee_name", employee.getEmployee_name());
        params.put("employee_salary", employee.getEmployee_salary());
        params.put("employee_age", String.valueOf(employee.getEmployee_age()));
        params.put("profile_image", employee.getProfile_image());
        return params;
    }

    // Respond Parsing

}

File: Task 2/Kotlin/app/src/main/java/dev/ogabek/kotlin/network/VolleyHttp.kt
package dev.ogabek.kotlin.network

import com.android.volley.Response
import com.android.volley.toolbox.StringRequest
import dev.ogabek.kotlin.MyApplication
import dev.ogabek.kotlin.helper.Logger
import dev.ogabek.kotlin.model.Employee
import org.json.JSONObject

class VolleyHttp {

    companion object {

        private val TAG = VolleyHttp::class.java.simpleName

        const val IS_TESTER = true
        private const val SERVER_DEVELOPMENT = "http://dummy.restapiexample.com/api/v1/"
        private const val SERVER_PRODUCTION = "http://dummy.restapiexample.com/api/v1/"

        fun server(url: String): String {
            return if (IS_TESTER) {
                SERVER_DEVELOPMENT + url
            } else {
                SERVER_PRODUCTION + url
            }
        }

        fun headers(): HashMap<String, String> {
            val header = HashMap<String, String>()
            header["Content-type"] = "application/json; charset=UTF-8"
            return header
        }

        const val API_GET_EMPLOYEES = "employees"
        const val API_GET_ONE_EMPLOYEE = "employee/"        //id
        const val API_POST_EMPLOYEES = "create"
        const val API_PUT_EMPLOYEES = "update/"        // id

        fun paramsEmpty(): HashMap<String, String> {
            return HashMap()
        }

        fun paramsPost(employee: Employee): HashMap<String, Any> {
            val params = HashMap<String, Any>()
            params["employee_name"] = employee.employee_name
            params["employee_salary"] = employee.employee_salary
            params["employee_age"] = employee.employee_age
            params["profile_image"] = employee.profile_image
            return params
        }

        fun paramsPut(employee: Employee): HashMap<String, Any> {
            val params = HashMap<String, Any>()
            params["id"] = employee.id
            params["employee_name"] = employee.employee_name
            params["employee_salary"] = employee.employee_salary
            params["employee_age"] = employee.employee_age
            params["profile_image"] = employee.profile_image
            return params
        }

        fun get(api: String, params: HashMap<String, String>, volleyHandler: VolleyHandler) {
            val request = object: StringRequest(
                Method.GET, server(api),
                Response.Listener { response ->
                    Logger.d(TAG, response)
                    volleyHandler.onSuccess(response)
                },
                Response.ErrorListener { error ->
                    Logger.e(TAG, error.message.toString())
                    volleyHandler.onSuccess(error.message.toString())
                }) {
                override fun getParams(): MutableMap<String, String> {
                    return params
                }
            }
            MyApplication.instance!!.addToRequestQueue(request)
        }

        fun post(api: String, body: HashMap<String, String>, volleyHandler: VolleyHandler) {
            val request = object: StringRequest(
                Method.POST, server(api),
                Response.Listener { response ->
                    Logger.d(TAG, response)
                    volleyHandler.onSuccess(response)
                },
                Response.ErrorListener { error ->
                    Logger.e(TAG, error.message.toString())
                    volleyHandler.onSuccess(error.message.toString())
                }) {
                override fun getHeaders(): MutableMap<String, String> {
                    return headers()
                }
                override fun getBody(): ByteArray {
                    return JSONObject(body as Map<*, *>).toString().toByteArray()
                }
            }
            MyApplication.instance!!.addToRequestQueue(request)
        }

        fun put(api: String, body: HashMap<String, String>, volleyHandler: VolleyHandler) {
            val request = object: StringRequest(
                Method.PUT, server(api),
                Response.Listener { response ->
                    Logger.d(TAG, response)
                    volleyHandler.onSuccess(response)
                },
                Response.ErrorListener { error ->
                    Logger.e(TAG, error.message.toString())
                    volleyHandler.onSuccess(error.message.toString())
                }) {
                override fun getHeaders(): MutableMap<String, String> {
                    return headers()
                }
                override fun getBody(): ByteArray {
                    return JSONObject(body as Map<*, *>).toString().toByteArray()
                }
            }
            MyApplication.instance!!.addToRequestQueue(request)
        }

        fun delete(api: String, volleyHandler: VolleyHandler) {
            val request = object: StringRequest(
                Method.DELETE, server(api),
                Response.Listener { response ->
                    Logger.d(TAG, response)
                    volleyHandler.onSuccess(response)
                },
                Response.ErrorListener { error ->
                    Logger.e(TAG, error.message.toString())
                    volleyHandler.onSuccess(error.message.toString())
                }) {

            }
            MyApplication.instance!!.addToRequestQueue(request)
        }

    }

}